File: server/RecipeManager.Application/Models/View/Recipe.cs
namespace RecipeManager.Application.Models.View;

public class Recipe
{
    public Guid? Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public ImageMeta? CoverImage { get; set; }
    public string Note { get; set; } = string.Empty;

    public List<IngredientGroup> IngredientGroups { get; set; } = new();
    public List<InstructionGroup> InstructionGroups { get; set; } = new();

    public string Category { get; set; } = string.Empty;
    public string Cuisine { get; set; } = string.Empty;
    public decimal Rating { get; set; } = decimal.Zero;
    
    public decimal Servings { get; set; }

    public RecipeDuration PreparationDuration { get; set; } = new("Preparation");
    public RecipeDuration CookingDuration { get; set; } = new("Cooking");
    public List<RecipeDuration> CustomDurations { get; set; } = new();
    
    public List<string> Tags { get; set; } = new();
    public string Slug { get; set; } = string.Empty;
}

File: website/types/recipe.ts
export type Recipe = {
  title: string;
  description: string;
  descriptionPlainText: string;
  descriptionSnippet: string;
  cuisine?: string;
  course?: string;
  note?: string;
  coverImage: Image;
  ingredientGroups: IngredientGroup[];
  instructionGroups: InstructionGroup[];
  servings?: number;
  servingsType?: string;
  preparationDuration?: number;
  cookingDuration?: number;
  customDurationName?: string;
  customDuration?: number;
  tags: string[];
  slug: string;
};

export type RecipePreview = {
  title: string;
  descriptionSnippet: string;
  coverImage: Image;
  cuisine?: string;
  course?: string;
  tags: string[];
  featuredTag?: string;
  preparationDuration?: number;
  cookingDuration?: number;
  customDurationName?: string;
  customDuration?: number;
  totalDuration?: string;
  slug: string;
  date_created: string;
  favourite?: boolean;
};

export type IngredientGroup = {
  name?: string;
  ingredients: Ingredient[];
};

export type Ingredient = {
  amount?: number;
  unit?: string;
  name: {
    singular: string;
    plural: string;
  };
  note?: string;
  inlineOnly?: boolean;
};

export type Image = {
  id: string;
  title: string;
  height: number;
  width: number;
  modifyDate: string;
};

export type InstructionGroup = {
  name?: string;
  instructions: Array<Instruction>;
};

export type Instruction = {
  text: string;
  image?: Image;
};
